File: tests/r_parity/60_sureg.py
"""StatsPAI seemingly unrelated regressions parity (Python side) -- Module 60.

Two-equation SUR with cross-correlated errors. sp.sureg's one-step FGLS
(Sigma = E'E / n from equation-by-equation OLS residuals, no df
correction) matches the Stata sureg default; the R side reproduces the
same convention via systemfit(method='SUR',
control=systemfit.control(methodResidCov='noDfCor')).
"""
from __future__ import annotations
import numpy as np, pandas as pd, statspai as sp
from _common import PARITY_SEED, ParityRecord, dump_csv, write_results

MODULE = "60_sureg"


def make_data(n=1000, seed=PARITY_SEED):
    rng = np.random.default_rng(seed)
    x1 = rng.normal(0, 1, n)
    x2 = rng.normal(0, 1, n)
    w = rng.normal(0, 1, n)
    e = rng.multivariate_normal([0.0, 0.0], [[1.0, 0.6], [0.6, 1.5]], size=n)
    y1 = 1.0 + 0.8 * x1 + 0.4 * w + e[:, 0]
    y2 = -0.5 + 0.6 * x2 - 0.3 * w + e[:, 1]
    return pd.DataFrame({"y1": y1, "y2": y2, "x1": x1, "x2": x2, "w": w})


def main():
    df = make_data()
    dump_csv(df, MODULE)
    res = sp.sureg({"eq1": ("y1", ["x1", "w"]),
                    "eq2": ("y2", ["x2", "w"])}, data=df, method="fgls")
    labels = ["eq1_intercept", "eq1_x1", "eq1_w",
              "eq2_intercept", "eq2_x2", "eq2_w"]
    rows = []
    for i, lab in enumerate(labels):
        rows.append(ParityRecord(MODULE, "py", f"beta_{lab}",
            estimate=float(res.params_all[i]),
            se=float(res.se_all[i]),
            n=int(res.n_obs)))
    write_results(MODULE, "py", rows,
                  extra={"method": "one-step FGLS", "sigma_divisor": "n"})


if __name__ == "__main__":
    main()

File: tests/r_parity/60_sureg.R
# StatsPAI seemingly unrelated regressions parity (R side) -- Module 60.
#
# systemfit SUR with methodResidCov='noDfCor' (Sigma divisor n) and
# maxiter=1 reproduces the one-step FGLS convention shared by
# sp.sureg(method='fgls') and the Stata sureg default.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) dirname(normalizePath(sub("^--file=", "", .file_arg[1]))) else getwd()
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({ library(systemfit) })

MODULE <- "60_sureg"
df <- read_csv_strict(MODULE)

eqs <- list(eq1 = y1 ~ x1 + w, eq2 = y2 ~ x2 + w)
ctl <- systemfit.control(methodResidCov = "noDfCor", maxiter = 1)
fit <- systemfit(eqs, method = "SUR", data = df, control = ctl)
co <- coef(fit); se <- sqrt(diag(vcov(fit)))

lab <- c(eq1_.Intercept. = "beta_eq1_intercept", eq1_x1 = "beta_eq1_x1", eq1_w = "beta_eq1_w",
         eq2_.Intercept. = "beta_eq2_intercept", eq2_x2 = "beta_eq2_x2", eq2_w = "beta_eq2_w")
nm <- names(co)
rows <- lapply(seq_along(co), function(i) {
  key <- gsub("[()]", ".", nm[i])
  parity_row(MODULE, unname(lab[key]), estimate = unname(co[i]), se = unname(se[i]), n = nrow(df))
})
write_results(MODULE, rows,
              extra = list(package = "systemfit::systemfit",
                           method = "SUR one-step FGLS",
                           residCov = "noDfCor (divisor n)"))
